### PR TITLE
Snippet translations + translatable fields

### LINKED ISSUE #6708: Step 6. Snippet translations
## Acceptance criteria 

> Note: Don't forget to reference [the document from Step 5 #6707](https://docs.google.com/spreadsheets/d/1AU7pxboZoBiVcbTpZO-VUMAhxZwx5mdEH_FBLfo2D-w/edit#gid=0)

- [ ] Create migrations for all snippets. [This document](https://docs.wagtail.io/en/stable/advanced_topics/i18n.html#making-snippets-with-existing-data-translatable) will be _very_ helpful. We need to bootstrap migrations for each model. 
  - [ ] Make sure all snippet models are translated. 
  - [ ] Make sure all snippet fields have `translatable_fields = [ .. ]` set on them 
  - [ ] Make sure tests still pass. May need to do some adjusting for new language headers in the `request` object, particularly around redirects. 
- [ ] Get Theo to review this (specifically, Theo) to make sure all the fields are the right translation type.

### PR DESCRIPTION
**Pull request by KalobTaulien**
Closes #6708
Related PRs/issues #6703 

## Acceptance criteria 

> Note: Don't forget to reference [the document from Step 5 #6707](https://docs.google.com/spreadsheets/d/1AU7pxboZoBiVcbTpZO-VUMAhxZwx5mdEH_FBLfo2D-w/edit#gid=0)

- [x] Create migrations for all snippets. [This document](https://docs.wagtail.io/en/stable/advanced_topics/i18n.html#making-snippets-with-existing-data-translatable) will be _very_ helpful. We need to bootstrap migrations for each model. 
  - [x] Make sure all snippet models are translated. 
  - [x] Make sure all snippet fields have `translatable_fields = [ .. ]` set on them 
  - [x] Make sure tests still pass. May need to do some adjusting for new language headers in the `request` object, particularly around redirects. 
- [ ] Get Theo to review this (specifically, Theo) to make sure all the fields are the right translation type.

### CODE REVIEW

**Comment by KalobTaulien** [comment]
Locally all the tests pass.

@Pomax I pinged you for code review on this. 
@TheoChevalier I pinged you for semi-code review but mostly to make sure I got all the fields right. Your doc was very helpful. I made some notes in the doc too and highlighted some lines for you to take a look at.

File: network-api/networkapi/news/models.py [hunk 1]
```diff
@@ -88,7 +89,7 @@ class News(models.Model):
 
     objects = NewsQuerySet.as_manager()
 
-    class Meta:
+    class Meta(TranslatableMixin.Meta):
```
Comment (Pomax) [inline network-api/networkapi/news/models.py:92]: Let's check if this is actually necessary or not. If we don't need to mark the Meta class as translatable, that saves some migration work. (not much, but "not doing anything we don't need to do is worth not doing" =D)

File: network-api/networkapi/wagtailpages/pagemodels/base.py [hunk 2]
```diff
@@ -2,13 +2,16 @@
...
-from wagtail.core.models import Page, Orderable as WagtailOrderable
+from wagtail.core.models import TranslatableMixin, Page, Orderable as WagtailOrderable
 from wagtail.core.fields import RichTextField
 from wagtail.images.edit_handlers import ImageChooserPanel
 from wagtail.snippets.models import register_snippet
 from wagtail.snippets.edit_handlers import SnippetChooserPanel
 from wagtail.admin.edit_handlers import PageChooserPanel
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/base.py:13]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/campaigns.py [hunk 3]
```diff
@@ -3,11 +3,14 @@
...
 
 from wagtail.admin.edit_handlers import FieldPanel, InlinePanel, StreamFieldPanel
-from wagtail.core.models import Page
+from wagtail.core.models import TranslatableMixin, Page
 from wagtail.core.fields import RichTextField
 from wagtail.snippets.edit_handlers import SnippetChooserPanel
 from wagtail.snippets.models import register_snippet
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/campaigns.py:12]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/dear_internet.py [hunk 4]
```diff
@@ -4,6 +4,9 @@
 from wagtail.core.models import Page
 from wagtail.core.fields import StreamField
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/dear_internet.py:8]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/index.py [hunk 5]
```diff
@@ -13,6 +13,9 @@
 from wagtail.core.models import Page
 from wagtail.contrib.routable_page.models import RoutablePageMixin, route
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/index.py:17]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/primary.py [hunk 6]
```diff
@@ -5,6 +5,9 @@
 from wagtail.core.fields import StreamField
 from wagtail.images.edit_handlers import ImageChooserPanel
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/primary.py:9]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/products.py [hunk 7]
```diff
@@ -17,12 +17,15 @@
...
 from wagtail.contrib.routable_page.models import RoutablePageMixin, route
-from wagtail.core.models import Orderable, Page
+from wagtail.core.models import TranslatableMixin, Orderable, Page
 from wagtail.images.edit_handlers import ImageChooserPanel
 from wagtail.search import index
 from wagtail.snippets.edit_handlers import SnippetChooserPanel
 from wagtail.snippets.models import register_snippet
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/products.py:27]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/publications/article.py [hunk 8]
```diff
@@ -8,6 +8,9 @@
 from wagtail.images.edit_handlers import ImageChooserPanel
 from wagtail.snippets.edit_handlers import SnippetChooserPanel
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/publications/article.py:12]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

File: network-api/networkapi/wagtailpages/pagemodels/publications/publication.py [hunk 9]
```diff
@@ -9,6 +9,9 @@
 from wagtail.images.edit_handlers import ImageChooserPanel
 from wagtail.snippets.edit_handlers import SnippetChooserPanel
 
+from wagtail_localize.fields import TranslatableField
+from wagtail_localize.fields import SynchronizedField
```
Comment (Pomax) [inline network-api/networkapi/wagtailpages/pagemodels/publications/publication.py:13]:
```suggestion
from wagtail_localize.fields import SynchronizedField, TranslatableField
```

**Review by Pomax — CHANGES_REQUESTED** [review]
some minor notes, and a question

**Comment by TheoChevalier** [comment]
I’ve made a few changes to the spreadsheet (all highlighted in orange), mostly answering your questions.
A few fields in snippets are not properly hooked (I’ve highlighted them in red in the Snippets sheet)

Also hit a few errors/have some questions:


### PNI General product:
Translating a general product raises:
Capture d’écran 2021-06-01 à 15 48 46

### Campaign page
Translating a campaign page that has a petition + donate snippets raises:
Capture d’écran 2021-06-01 à 14 15 12


### Youtube Regrets
`Regret stories` are missing. Is this because it’s a block and not part of the model?

### Article pages:
For Text and ID https://docs.google.com/spreadsheets/d/1AU7pxboZoBiVcbTpZO-VUMAhxZwx5mdEH_FBLfo2D-w/edit#gid=0&range=A43 these are fields I’ve found at the bottom of article pages, but I haven’t been able to find them in the model code
Capture d’écran 2021-06-01 à 16 04 55

### Homepage
Here too, a lot of the content is not localized, is this because these are blocks and it will come later?
Also it seems every page picker field need to be `Translatable`, otherwise links will point to /en/…
This affects all page pickers in News you can use, Spotlight, Take action, etc

**Comment by KalobTaulien** [comment]
Hey @TheoChevalier Thank you for digging into this! 

1. Ahh there was a typo in the PNI products. Glad localize found that. Should be fixed now. 
2. Not sure how but I missed the MozfestHomepage, so Ive added that in. 
3. I made some changes you suggested from your doc as well. 
4. The Homepage [has translatable_fields here](https://github.com/mozilla/foundation.mozilla.org/pull/6777/files#diff-acb6632b9522d55f99ba1616f58beda704393419a37a5674e86d2f7e12d825f6R771-R791) its just not easy to see in the Changed Files tab. Let me know if I'm missing any. 
5. Campaign pages should translate now. But let me know if it doesn't and how you got to that point. They worked locally for me.

**Comment by KalobTaulien** [comment]
We've run into an issue with Wagtail Footnotes. The Footnote model and the localize Translate model both have `uuid = ...` in them causing a conflict. I pinged Karl about this since I'm unsure how to proceed with this properly.

**Comment by TheoChevalier** [comment]
> Hey @TheoChevalier Thank you for digging into this!
> 
>     1. Ahh there was a typo in the PNI products. Glad localize found that. Should be fixed now.

Yup, thanks!
> 
>     2. Not sure how but I missed the MozfestHomepage, so Ive added that in.

Hm, I’m getting this error when translating it
Capture d’écran 2021-06-02 à 20 02 38


>     3. I made some changes you suggested from your doc as well.
> 
>     4. The Homepage [has translatable_fields here](https://github.com/mozilla/foundation.mozilla.org/pull/6777/files#diff-acb6632b9522d55f99ba1616f58beda704393419a37a5674e86d2f7e12d825f6R771-R791) its just not easy to see in the Changed Files tab. Let me know if I'm missing any.


> 
>     5. Campaign pages should translate now. But let me know if it doesn't and how you got to that point. They worked locally for me.

Thanks! I’m able to submit now, but I end up with an empty French page… Do you see that as well?
I just clicked "Translate this page" on the first campaign page I saw into the two locales I’ve set up (German & French), then edited the French translation
Capture d’écran 2021-06-02 à 19 38 30


Another thing I noticed, strings from the Promote tab are missing from all pages (SEO title and description), I suspect these are a bit special?

**Comment by KalobTaulien** [comment]
Hey @TheoChevalier I'll take a look at the campaign pages acting up in a bit. Thanks for catching that! 

I opened an issue here #6790 regarding the footnotes issue. I've disabled footnote translations for the time being in this PR so we can land this asap.

**Review by Pomax — CHANGES_REQUESTED** [review]
so it looks like snippets are not actually working, let's get some time with Karl to figure this one out.

File: network-api/networkapi/wagtailpages/migrations/0012_bootstrap_migrations.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('wagtailpages', '0011_auto_20210531_1735'),
+    ]
+
+    operations = [
+        BootstrapTranslatableModel('wagtailpages.DonationModal'),
+        BootstrapTranslatableModel('wagtailpages.FocusArea'),
+        BootstrapTranslatableModel('wagtailpages.FocusArea'),
```
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/migrations/0012_bootstrap_migrations.py:17]: Manual migration. Copy and paste got me here. Will remove, thanks for finding this!

File: network-api/networkapi/mozfest/models.py [hunk 11]
```diff
@@ -167,10 +169,25 @@ class MozfestHomepage(MozfestPrimaryPage):
...
+        TranslatableField('banner_heading'),
+        TranslatableField('banner_guide_text'),
+        SynchronizedField('banner_video_url'),
+        TranslatableField('title'),
+        TranslatableField('search_description'),
+        TranslatableField('search_image'),
+        TranslatableField('signup'),
+        TranslatableField('body'),
+        # TODO: Add in footnote support when wagtail-footnotes and wagtail-localize can work together.
+        # TranslatableField('footnotes'),
```
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply network-api/networkapi/mozfest/models.py:196]: We do! And good call. Will do that now.

File: network-api/networkapi/wagtailpages/pagemodels/products.py [hunk 12]
```diff
@@ -698,6 +708,52 @@ def get_voting_json(self):
...
+        TranslatableField('strong_password_helptext'),
+        SynchronizedField('manage_vulnerabilities'),
+        TranslatableField('manage_vulnerabilities_helptext'),
+        SynchronizedField('privacy_policy'),
+        TranslatableField('privacy_policy_helptext'),
+        SynchronizedField('phone_number'),
+        SynchronizedField('live_chat'),
+        SynchronizedField('email'),
+        SynchronizedField('twitter'),
+    ]
```
[the remark this answers is not in the document]
Comment (KalobTaulien) [reply network-api/networkapi/wagtailpages/pagemodels/products.py]: This is a good question for Karl tomorrow. I also don't like looking at this. It feels very.. tedious.

**Review by Pomax — CHANGES_REQUESTED** [review]
looks like snippets at least work now, but we'll need to comb the code on Monday to make sure we caught all of them.

**Review by Pomax — APPROVED** [review]
Let's get this landed and then deal with the followup as new work.

**Comment by KalobTaulien** [comment]
Ticket made for the campaign page problem. #6851

**Pull request merged**